numeric.c

```diff
@@ -2170,36 +2170,82 @@ flo_ndigits(int argc, VALUE *argv)
+ *  as specified by `ndigits`,
+ *  which must be an
+ *  [integer-convertible object](rdoc-ref:implicit_conversion.rdoc@Integer-Convertible+Objects).
+ *
+ *  When `self` is zero,
+ *  returns a zero value:
+ *  a float if `ndigits` is positive,
+ *  an integer otherwise:
+ *
+ *  ```
+ *  f = 0.0      # => 0.0
+ *  f.floor(20)  # => 0.0
+ *  f.floor(0)   # => 0
+ *  f.floor(-20) # => 0
+ *  ```
  *
- *  When +ndigits+ is positive, returns a float with +ndigits+
+ *  When `self` is non-zero and `ndigits` is positive, returns a float with `ndigits`
  *  digits after the decimal point (as available):
  *
- *    f = 12345.6789
- *    f.floor(1) # => 12345.6
- *    f.floor(3) # => 12345.678
- *    f = -12345.6789
- *    f.floor(1) # => -12345.7
- *    f.floor(3) # => -12345.679
+ *  ```
+ *  f = 12345.6789
+ *  f.floor(1)  # => 12345.6
+ *  f.floor(3)  # => 12345.678
+ *  f.floor(30) # => 12345.6789
+ *  f = -12345.6789
+ *  f.floor(1)  # => -12345.7
+ *  f.floor(3)  # => -12345.679
+ *  f.floor(30) # => -12345.6789
+ *  ```
  *
- *  When +ndigits+ is non-positive, returns an integer with at least
- *  <code>ndigits.abs</code> trailing zeros:
+ *  When `self` is non-zero and `ndigits` is non-positive,
+ *  returns an integer value based on a computed granularity:
  *
- *    f = 12345.6789
- *    f.floor(0)  # => 12345
- *    f.floor(-3) # => 12000
- *    f = -12345.6789
- *    f.floor(0)  # => -12346
- *    f.floor(-3) # => -13000
+ *  - The granularity is `10 ** ndigits.abs`.
+ *  - The returned value is the largest multiple of the granularity
+ *    that is less than or equal to `self`.
+ *
+ *  Examples with positive `self`:
+ *
+ *  | ndigits | Granularity | 12345.6789.floor(ndigits) |
+ *  |--------:|------------:|--------------------------:|
+ *  |       0 |           1 |                     12345 |
+ *  |      -1 |          10 |                     12340 |
+ *  |      -2 |         100 |                     12300 |
+ *  |      -3 |        1000 |                     12000 |
+ *  |      -4 |       10000 |                     10000 |
+ *  |      -5 |      100000 |                         0 |
+ *
+ *  Examples with negative `self`:
+ *
+ *  | ndigits | Granularity | -12345.6789.floor(ndigits) |
+ *  |--------:|------------:|---------------------------:|
+ *  |       0 |           1 |                     -12346 |
+ *  |      -1 |          10 |                     -12350 |
+ *  |      -2 |         100 |                     -12400 |
+ *  |      -3 |        1000 |                     -13000 |
+ *  |      -4 |       10000 |                     -20000 |
+ *  |      -5 |      100000 |                    -100000 |
+ *  |      -6 |     1000000 |                   -1000000 |
+ *  |      -7 |    10000000 |                  -10000000 |
+ *  |      -8 |   100000000 |                 -100000000 |
+ *  |      -9 |  1000000000 |                -1000000000 |
+ *  |     -10 | 10000000000 |                          0 |
```

Comment:
I'm not sure why the result is 0 here, is it not `-10000000000`?

```ruby
-12345.6789.floor(-10) # => -10000000000
```

Reply:
Not on my machine:
```
-12345.6789.floor(-10)
# => 0
-12345.6789.floor(-20)
# => 0
```

```
irb --version
irb 1.11.0 (2023-12-19)
```

Reply:
That's odd. What's your Ruby version (can you paste the output of `ruby -v`)? Also, what's the output of your `-12345.floor(-10)`? It should be `-10000000000`.

Reply:
```
ruby -v
ruby 3.3.0 (2023-12-25 revision 5124f9ac75) [x64-mingw-ucrt]
```

```
irb
-12345.floor(-10)
# => 0
```

Reply:
I also see the difference between the two.  I'll be happy to confirm and put up the Linux results if that seems best.

Reply:
Oh I see what's going on. When the rounding will overflow, it returns 0 instead. Windows' long is 4 bytes whereas on macOS/Linux it's 8 bytes. I can reproduce the result with a larger magnitude `ndigits`.

```ruby
-12345.6789.floor(-100) # => 0
-12345.floor(-100) # => 0
12344.ceil(-100) # => 0
```

I'm not sure if this is intended behaviour. If it's not intended behaviour, then we shouldn't document it. If it is, then we should have a separate section to describe it.

@nobu do you know if this is intended behaviour?

Reply:
I have removed the questionable lines from the table.

Reply:
I opened a PR that will fix this issue in https://github.com/ruby/ruby/pull/11257.